pkg/integrations/v2/app_o11y_receiver/sourcemaps/sourcemaps.go

```diff
@@ -0,0 +1,331 @@
+			return nil, "", err
+		}
+		relative, err := url.Parse(sourceMapURL)
+		if err != nil {
+			level.Debug(store.l).Log("msg", "failed to parse source map url", "url", sourceURL, "sourceMapURL", sourceMapURL, "err", err)
+			return nil, "", err
+		}
+		resolvedSourceMapURL = base.ResolveReference(relative).String()
+		level.Debug(store.l).Log("msg", "resolved absolute soure map url", "url", sourceURL, "sourceMapURL", resolvedSourceMapURL)
+	}
+	level.Debug(store.l).Log("msg", "attempting to download sourcemap file", "url", resolvedSourceMapURL)
+	result, err = store.downloadFileContents(resolvedSourceMapURL)
+	if err != nil {
+		level.Debug(store.l).Log("failed to download source map file", "url", resolvedSourceMapURL, "err", err)
+		return nil, "", err
+	}
+	return result, resolvedSourceMapURL, nil
+}
+
+func (store *RealSourceMapStore) getSourceMapFromFileSystem(sourceURL string, release string, fileconf config.SourceMapFileLocation) (content []byte, sourceMapURL string, err error) {
+	if len(sourceURL) == 0 || !strings.HasPrefix(sourceURL, fileconf.MinifiedPathPrefix) || strings.HasSuffix(sourceURL, "/") {
+		return nil, "", nil
+	}
+	pathParts := []string{strings.Replace(fileconf.Path, "{RELEASE}", cleanForFilePath(release), 1)}
+	for _, part := range strings.Split(strings.TrimPrefix(sourceURL, fileconf.MinifiedPathPrefix), "/") {
+		if len(part) > 0 && part != "." && part != ".." {
+			pathParts = append(pathParts, part)
+		}
+	}
+	mapFilePath := filepath.Join(pathParts...) + ".map"
+
+	if _, err := store.fileService.Stat(mapFilePath); err != nil {
+		store.metrics.fileReads.WithLabelValues(getOrigin(sourceURL), "not_found").Inc()
+		level.Debug(store.l).Log("msg", "sourcemap not found on filesystem", "url", sourceURL, "file_path", mapFilePath)
+		return nil, "", nil
+	}
+	level.Debug(store.l).Log("msg", "sourcemap found on filesystem", "url", mapFilePath, "file_path", mapFilePath)
+
+	content, err = store.fileService.ReadFile(mapFilePath)
+	if err != nil {
+		store.metrics.fileReads.WithLabelValues(getOrigin(sourceURL), "error").Inc()
+	} else {
+		store.metrics.fileReads.WithLabelValues(getOrigin(sourceURL), "ok").Inc()
+	}
+	return content, sourceURL, err
+}
+
+func (store *RealSourceMapStore) getSourceMapContent(sourceURL string, release string) (content []byte, sourceMapURL string, err error) {
+	//attempt to find in fs
+	for _, fileconf := range store.config.FileSystem {
+		content, sourceMapURL, err = store.getSourceMapFromFileSystem(sourceURL, release, fileconf)
+		if content != nil || err != nil {
+			return content, sourceMapURL, err
+		}
+	}
+
+	//attempt to download
+	if strings.HasPrefix(sourceURL, "http") && utils.URLMatchesOrigins(sourceURL, store.config.DownloadFromOrigins) {
+		return store.downloadSourceMapContent(sourceURL)
+	}
+	return nil, "", nil
+}
+
+func (store *RealSourceMapStore) getSourceMap(sourceURL string, release string) (*sourceMap, error) {
+	store.Lock()
```

Comment:
I am not 100% with lazy loading the source maps and I will share my thoughts. First of all, this will make the first request from each instance to be delayed a bit. I would probably go for eager loading in this case on startup.

Reply:
That being said the lock will delay all concurrent requests while the app is loading

Reply:
I have thought about these issues, here's my thinking.

Main argument:
* This is the simplest approach with least effort and complexity. I suspect that it might be _good enough_, so would rather give it a try and collect some data before optimizing

Supporting arguments:
* Slight delay does not really impact antyhing, frontend is not relying on quick response
* There's unlikely to be *many* sourcemaps to load in a typical project. I'd say Grafana is a fairly big frontend, and it builds to only dozens of source bundles
* In case of downloads, sourcemap locations are not known ahead of time so cannot be eagerly loaded
* Loading from file is fast, and we don't expect to have to load many files

Reply:
Lets go with that then and keep an eye on it